src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -46832,6 +46832,1397 @@ class CL1
+        [Theory]
+        [InlineData("explicit")]
+        [InlineData("implicit")]
+        public void NullCoalescing_CondAccess_ExplicitUserDefinedConv_03(string conversionKind)
+        {
+            var source = @"
+struct B { }
+struct C
+{
+    public static " + conversionKind + @" operator B(C c) => new B();
+
+    void M1(C? c1, object? x)
+    {
+        B b = (B?)c1?.M1(x = 0) ?? c1!.Value.M2(x = 0);
+        x.ToString();
+    }
+
+    C M1(object obj) { return this; }
+    B M2(object obj) { return new B(); }
+}
+";
+            CreateNullableCompilation(source).VerifyDiagnostics();
+        }
+
+        ///<remarks>Ported from <see cref="FlowTests.NullCoalescing_CondAccess_UserDefinedConv_04"/>.</remarks>
+        [Theory]
+        [InlineData("explicit")]
+        [InlineData("implicit")]
+        public void NullCoalescing_CondAccess_ExplicitUserDefinedConv_04(string conversionKind)
+        {
+            var source = @"
+struct B { }
+struct C
+{
+    public static " + conversionKind + @" operator B?(C c) => null;
+
+    void M1(C? c1, object? x)
+    {
+        B? b = (B?)c1?.M1(x = 0) ?? c1!.Value.M2(x = 0);
+        x.ToString();
+    }
+
+    C M1(object obj) { return this; }
+    B? M2(object obj) { return new B(); }
+}
+";
+            CreateNullableCompilation(source).VerifyDiagnostics();
+        }
+
+        ///<remarks>Ported from <see cref="FlowTests.NullCoalescing_CondAccess_UserDefinedConv_05"/>.</remarks>
+        [Theory]
+        [InlineData("explicit")]
+        [InlineData("implicit")]
+        public void NullCoalescing_CondAccess_ExplicitUserDefinedConv_05_01(string conversionKind)
+        {
+            var source = @"
+struct B
+{
+    public static " + conversionKind + @" operator B(C? c) => default;
+}
+
+class C
+{
+    static void M1(C c1, object? x)
+    {
+        B b = (B?)c1?.M1(x = 0) ?? c1!.M2(x = 0);
+        x.ToString(); // 1
+    }
+
+    C M1(object obj) { return this; }
+    B M2(object obj) { return default; }
+}
+";
+            CreateNullableCompilation(source).VerifyDiagnostics(
+                // (12,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 9)
+                );
+        }
+
+        ///<remarks>Ported from <see cref="FlowTests.NullCoalescing_CondAccess_UserDefinedConv_05"/>.</remarks>
+        [Theory]
+        [InlineData("explicit")]
+        [InlineData("implicit")]
+        public void NullCoalescing_CondAccess_ExplicitUserDefinedConv_05_02(string conversionKind)
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+struct B
+{
+    public static " + conversionKind + @" operator B([DisallowNull] C? c) => default;
+}
+
+class C
+{
+    static void M1(C c1, object? x)
+    {
+        B b = (B?)c1?.M1(x = 0) ?? c1!.M2(x = 0); // 1
```

Comment:
This one feels a bit wonky to me. My initial reaction is that it should warn for this case.

Reply:
We do warn for this case--just on the conversion instead of on usage of the updated variables. :)

The principle behind the behavior you're seeing is that we can propagate out "state when not null" from a user-defined conversion *unless* it can return non-null output for a null input. The rule I implemented based on this is that we can propagate out "state when not null" under the following conditions:

1. either the parameter is non-nullable, or
2. the return is nullable and has a `[return: NotNullIfNotNull]` referencing the parameter.

The first rule means in practice that the user-defined conversion needs to have a non-nullable value type parameter and be lifted. Otherwise we either gave a warning on the conversion or there was a suppression on the operand.

Alternatively, we could adjust the rule to say we can propagate out "state when not null" when:

1. the parameter is of a non-nullable value type, or
2. the return is nullable and has a `[return: NotNullIfNotNull]` referencing the parameter.

I'm not extremely attached to it going either way.